---
title: "0086 - [M220JS]Ticket 4.3: Handling Errors"
tags: ["MongoDB University"]
album: "M220JS: MongoDB for JavaScript Developers"
---

## 问题

**Task**

For this ticket, you'll be required to make the API more robust by handling exceptions. Specifically, what would happen should an incorrectly formatted `_id` be passed to **getMovieByID()** in `moviesDAO.js`?

In this case, an `InvalidId` error will be thrown to **getMovieByID()**, but the method does not need to _return_ this error. Instead, if this error is thrown, the method should return `null`.

A try/catch block is already included for you in **getMovieByID()**. Use the variable e to figure out if the `InvalidId` error is being thrown, and then return `null` in this case.

**Hint**

When the error `e` is caught, it has type `Error`. You might want to convert this to a string.

---

**Testing and Running the Application**

You can run the unit tests for this ticket by running:

```bash
npm test -t error-handling
```

Once the unit tests are passing, run the application with:

```
npm start
```

Now proceed to the status page to run the full suite of integration tests and get your validation code.

After passing the relevant tests, what is the validation code for **Error Handling**?

<!--more-->

## 解析

题目要求只在`_id`参数出错时返回`null`，其他异常情况下应仍然抛出异常：

```js
  static async getMovieByID(id) {
    try {
      // ...
    } catch (e) {
      if (e.toString().includes("Error: Argument passed in must be a single String of 12 bytes or a string of 24 hex characters")) {
        return null
      } else {
        throw e
      }
    }
  }
}
```

## 答案

5ae9b76a703c7c603202ef22
